<?xml version = "1.0"?>
<robot name = "scara" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <!-- Robot arm dimension -->
  <xacro:property name="PI" value="3.1415926535897931"/>

  <material name="black">
    <color rgba="0.0 0.0 0.0 1.0"/>
  </material>

  <material name="blue">
    <color rgba="0.0 0.0 0.8 1.0"/>
  </material>

  <material name="green">
    <color rgba="0.0 0.8 0.0 1.0"/>
  </material>

  <material name="orange">
    <color rgba="1 0.4 0.1 1.0"/>
  </material>

  <material name="brown">
    <color rgba="1 0.87 0.84 1.0"/>
  </material>

  <material name="grey">
    <color rgba="0.2 0.2 0.2 1.0"/>
  </material>


  <material name="red">
    <color rgba="0.8 0.0 0.0 1.0"/>
  </material>

  <material name="white">
    <color rgba="1.0 1.0 1.0 1.0"/>
  </material>

  <!-- For gazebo colors -->
  <gazebo reference="base_link">
    <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="revolute1">
    <material>Gazebo/Red</material>
  </gazebo>
  <gazebo reference="link1">
    <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="revolute2">
    <material>Gazebo/Red</material>
  </gazebo>
  <gazebo reference="link2">
    <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="prismatic3">
    <material>Gazebo/Blue</material>
  </gazebo>
  <gazebo reference="link3">
    <material>Gazebo/Black</material>
  </gazebo>



  <!-- Used for fixing robot to Gazebo 'base_link' -->
  <link name="world"/>

  <!-- Base Link -->
  <link name = "base_link">
    <visual>
      <geometry>
        <box size = "0.12 0.12 0.05"/>
      </geometry>
      <material name = "blue"/>
      <origin xyz = "0 0 0.025"/>
    </visual>
    <inertial>
      <mass value="1" />
      <origin xyz="0 0 0.0" rpy="0 0 0" />
      <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001" />
    </inertial>
    <collision>
      <geometry>
        <box size = "0.12 0.12 0.05"/>
      </geometry>
      <material name = "blue"/>
      <origin xyz = "0 0 0.025"/>
    </collision>
  </link>

  <joint name="base2world" type="fixed">
    <parent link="world"/>
    <child link="base_link"/>
  </joint>

  <!-- revolute1 1 -->
  <link name = "revolute1">
    <visual>
      <geometry>
        <cylinder length = "0.45" radius = "0.05"/>
      </geometry>
      <material name = "green"/>
      <origin xyz = "0 0 0.225"/>
    </visual>
    <inertial>
      <mass value="0.5" />
      <origin xyz="0 0 0.0" rpy="0 0 0" />
      <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001" />
    </inertial>
    <collision>
      <geometry>
        <cylinder length = "0.45" radius = "0.05"/>
      </geometry>
      <material name = "green"/>
      <origin xyz = "0 0 0.225"/>
    </collision>
  </link>

  <joint name = "joint1" type="continuous">
    <parent link = "base_link"/>
    <child link = "revolute1"/>
    <origin xyz = "0 0 0.05"/>
    <limit effort="1000.0" lower="-1.57" upper="1.57" velocity="0.5"/>
    <axis xyz = "0 0 1"/>
    <dynamics damping="0.2" friction="0.1" />
  </joint>


  <!-- link 1 -->
  <link name = "link1">
    <visual>
      <geometry>
        <box size = "0.1 0.1 0.425"/>
      </geometry>
      <material name = "blue"/>
      <origin rpy = "0 1.571 0" xyz = "0.2125 0 0"/>
    </visual>
    <inertial>
      <mass value="0.5" />
      <origin xyz="0 0 0.0" rpy="0 0 0" />
      <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001" />
    </inertial>
    <collision>
      <geometry>
        <box size = "0.1 0.1 0.425"/>
      </geometry>
      <material name = "blue"/>
      <origin rpy = "0 1.571 0" xyz = "0.2125 0 0"/>
    </collision>
  </link>

  <joint name = "link12revolute1" type="fixed">
    <parent link = "revolute1"/>
    <child link = "link1"/>
    <origin xyz = "0 0 0.4"/>
  </joint>
     
  <!-- revolute 2 -->
  <link name = "revolute2">
    <visual>
      <geometry>
        <cylinder length = "0.2" radius = "0.05"/>
      </geometry>
      <material name = "green"/>
      <origin rpy = "0 0 0" xyz = "0 0 0"/>
    </visual>
    <inertial>
      <mass value="0.5" />
      <origin xyz="0 0 0.0" rpy="0 0 0" />
      <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001" />
    </inertial>
    <collision>
      <geometry>
        <cylinder length = "0.2" radius = "0.05"/>
      </geometry>
      <material name = "green"/>
      <origin rpy = "0 0 0" xyz = "0 0 0"/>
    </collision>
  </link>

  <joint name = "joint2" type="continuous">
    <parent link = "link1"/>
    <child link = "revolute2"/>
    <origin xyz = "0.425 0 0.05"/>
    <limit effort="1000.0" lower="-1.57" upper="1.57" velocity="0.5"/>
    <axis xyz = "0 0 1"/>
    <dynamics damping="0.2" friction="0.1" />
  </joint>

  <!-- link 2 -->
  <link name = "link2">
    <visual>
      <geometry>
        <box size = "0.1 0.1 0.345"/>
      </geometry>
      <material name = "blue"/>
      <origin rpy = "0 1.571 0" xyz = "0.1725 0 0"/>
    </visual>
    <inertial>
      <mass value="0.5" />
      <origin xyz="0 0 0.0" rpy="0 0 0" />
      <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001" />
    </inertial>
    <collision>
      <geometry>
        <box size = "0.1 0.1 0.345"/>
      </geometry>
      <material name = "blue"/>
      <origin rpy = "0 1.571 0" xyz = "0.1725 0 0"/>
    </collision>
  </link>

  <joint name = "link22revolute2" type="fixed">
    <parent link = "revolute2"/>
    <child link = "link2"/>
    <origin xyz = "0 0 0.05"/>
  </joint>

  <!-- prismatic 3 -->
  <link name = "prismatic3">
    <visual>
      <geometry>
        <cylinder length = "0.15" radius = "0.07"/>
      </geometry>
      <material name = "orange"/>
      <origin xyz = "0 0 0"/>
    </visual>
    <inertial>
      <mass value="0.5" />
      <origin xyz="0 0 0.0" rpy="0 0 0" />
      <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001" />
    </inertial>
    <collision>
      <geometry>
        <cylinder length = "0.15" radius = "0.07"/>
      </geometry>
      <material name = "orange"/>
      <origin xyz = "0 0 0"/>
    </collision>
  </link>
  
  <joint name = "prismatic32link2" type="fixed">
    <parent link = "link2"/>
    <child link = "prismatic3"/>
    <origin rpy = "3.14 0 0" xyz = "0.345 0 0"/>
  </joint> 

  <!-- link 3 -->
  <link name = "link3">
    <visual>
      <geometry>
        <cylinder length = "0.5" radius = "0.03"/>
      </geometry>
      <material name = "blue"/>
      <origin xyz = "0 0 -0.25"/>
    </visual>
    <inertial>
      <mass value="0.2" />
      <origin xyz="0 0 0.0" rpy="0 0 0" />
      <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001" />
    </inertial>
    <collision>
      <geometry>
        <cylinder length = "0.5" radius = "0.03"/>
      </geometry>
      <material name = "blue"/>
      <origin xyz = "0 0 -0.25"/>
    </collision>
  </link>
  
  <joint name = "joint3" type="prismatic">
    <parent link = "prismatic3"/>
    <child link = "link3"/>
    <origin xyz = "0 0 0.11"/>
    <limit effort="1000" lower="-0.00001" upper="0.3" velocity="0.1"/> 
    <axis xyz = "0 0 1"/> 
    <dynamics damping="0.4" friction="5" />
  </joint>

  <!-- Controller -->
  <transmission name="tran1">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint1">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor1">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="tran2">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint2">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor2">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="tran3">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint3">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor3">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  
  <joint name="camera_joint" type="fixed">
    <axis xyz="0 1 0" />
    <origin xyz="0 0 1.6" rpy="0 1.570796327 0"/>
    <parent link="world"/>
    <child link="camera_link"/>
  </joint>

  <!-- Camera -->
  <link name="camera_link">
        <visual>
      <origin xyz="0.0 0.0 0" rpy="0 0 0"/>
      <geometry>
        <box size="0.01 0.01 0.01"/>
      </geometry>
    </visual>

    <inertial>
      <mass value="1e-5" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>

     <!-- camera -->
  <gazebo reference="camera_link">
    <sensor type="camera" name="camera">
      <update_rate>30.0</update_rate>
      <camera name="head">
        <horizontal_fov>1.3962634</horizontal_fov>
        <image>
          <width>800</width>
          <height>800</height>
          <format>R8G8B8</format>
        </image>
        <clip>
          <near>0.02</near>
          <far>300</far>
        </clip>
        <noise>
          <type>gaussian</type>
          <!-- Noise is sampled independently per pixel on each frame.
               That pixel's noise value is added to each of its color
               channels, which at that point lie in the range [0,1]. -->
          <mean>0.0</mean>
          <stddev>0.007</stddev>
        </noise>
      </camera>
      <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
        <alwaysOn>true</alwaysOn>
        <updateRate>0.0</updateRate>
        <cameraName>camera</cameraName>
        <imageTopicName>image_raw</imageTopicName>
        <cameraInfoTopicName>camera_info</cameraInfoTopicName>
        <frameName>camera_link</frameName>
        <hackBaseline>0.07</hackBaseline>
        <distortionK1>0.0</distortionK1>
        <distortionK2>0.0</distortionK2>
        <distortionK3>0.0</distortionK3>
        <distortionT1>0.0</distortionT1>
        <distortionT2>0.0</distortionT2>
      </plugin>
    </sensor>
  </gazebo>

  <gazebo>
    <plugin name="control" filename="libgazebo_ros_control.so"/>
  </gazebo>

</robot>